### PR TITLE
blog: author name clickable, description and posts by the author visible

### LINKED ISSUE #1120: blog: make author name clickable, show description and posts by the author

### PR DESCRIPTION
**Pull request by utkarshsingh99**
Fix #1120 

- [X] Please read the guidelines in the [Contributing to the Documentation](https://dvc.org/doc/user-guide/contributing/docs) list if you make any substantial changes to the documentation or JS engine.

- [X] 🐛 Please make sure to mention `Fix #issue` (if applicable) in the description of the PR. This causes GitHub to close it automatically when the PR is merged.

- [X] Please choose to [allow us](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to edit your branch when creating the PR.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

**Built a new Author Page that looks like this**:
snapshot 1
snapshot 2
The link for every author is their `sourcePath` except ~~.md~~
If the sourcePath is: _authors/dmitry_petrov.md_, the link would be: **authors/dmitry_petrov**.

> make author name clickable

Resolved: The link on every blog post for the author's social media handle has been replaced with the link for their author page. 

> show description and posts by the author

Done. Unfortunately, the authors node in graphql does not have the description field unlike the markdown files of authors. Thus, I was not able to add a great description and had to just add their social media links.

For the future, as #1349 mentions, we can add more links in `Blog/Author/index.tsx` by passing the required props.

### CODE REVIEW

File: src/gatsby/models/authors/createPages.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+          next,
+          previous,
+          id
+        },
+        path: sourcePath.slice(0, sourcePath.indexOf('.'))
+      })
+    })
+  )
+
+  return Promise.all([authorPagesPromise])
```
Comment (rogermparent) [inline src/gatsby/models/authors/createPages.js:57]: You can just return `authorPagesPromise` with no wrapper here, `Promise.all` is for if we want to wait on multiple Promises happening at the same time.

File: content/docs/command-reference/import.md [hunk 2]
```diff
@@ -137,6 +137,39 @@ Several of the values above are pulled from the original stage file
...
+      url: git@github.com:GSA/data
+      rev_lock: af6a1feb542dc05b4d3e9c80deb50e6596876e5f
+outs:
+  - md5: 7e6de779a1ab286745c808f291d2d671
+    path: it-standards.csv
+    cache: true
+    metric: false
+    persist: false
+```
+
```
Comment (rogermparent) [inline content/docs/command-reference/import.md:172]: Someone else will have to make sure this is accurate to DVC and formatted how we want it.

**Comment by rogermparent** [comment]
I just made the change that generates a `slug` at build time and uses that. We may also want to link this page in other places where Author names show up, like the Post index.

**Comment by rogermparent** [comment]
> Done. Unfortunately, the authors node in graphql does not have the description field unlike the markdown files of authors. Thus, I was not able to add a great description and had to just add their social media links.

I can add a description field here, that's no problem!

Actually, what *do* we want for descriptions? We could have something like `longDescription` on blog posts which is externally precompiled markdown in the frontmatter, or we could use the Author's Markdown body content in whole or in part.

**Comment by rogermparent** [comment]
Also, I'm open to rolling #1349 into this and adding the multiple links support. It'd make more sense to introduce the schema change with a feature that uses it.

**Comment by rogermparent** [comment]
GitHub didn't pick up on the deploy I triggered, but it was sucessful and anyone who's interested in seeing the current Author page can see it [here](https://dvc-landing-author-page-afpkg5.herokuapp.com/authors/elle_obrien)

**Comment by rogermparent** [comment]
I added a quick fix to an issue I started to run into where deploy branches have problems re-using their cache. I need to diagnose this more, but since it was so small I added it here to get the deploy rolling.

File: content/docs/command-reference/import.md (hunk 2 above)
Comment (shcheklein) [reply content/docs/command-reference/import.md:172]: @utkarshsingh99 it looks like it is from a wrong branch?

**Comment by shcheklein** [comment]
Things we need to do before we merge/move forward with this:

- [ ] Design work for the author page header. It looks quite ugly at the moment.
- [ ] Include author descriptions.
- [ ] Authors should be clickable on the main page as well as blog page.
- [ ] We need to show some summary, or even Twitter/Linkedin/Email icons near the author or when you hover over it on the blog post page
- [ ] We need pagination for the author page.

**Comment by rogermparent** [comment]
> Design work for the author page header. It looks quite ugly at the moment.

Agreed

> Include author descriptions.

I have the page adding the body via `html` currently, which is one possibly way we can handle descriptions. I'm open to changing this to another method if we think it's easier for editors.

>  Authors should be clickable on the main page as well as blog page.

Agreed, and simple enough

> We need to show some summary, or even Twitter/Linkedin/Email icons near the author or when you hover over it on the blog post page

I think the most efficient course of action here would be to roll #1349 into this PR, as it will both affect the design of this page and serve as a source for Author social icons to be added throughout the site.

> We need pagination for the author page.

Agreed, and I think the existing pagination logic has needed a touch-up for a while anyway so I'll get to look at it when I re-use it here.

@utkarshsingh99 How would you like to divide the work left on this repo? I'm happy to take as much or as little as you'd like.
I'll start with doing #1349 since I started the issue, and then tackle pagination since that involves the Gatsby APIs heavily. Would you like to iterate on the design of the template? It would help a lot having that done at the same time as these schema changes.

And, of course, if you want or need to drop this I can pick it up. We've gone a little out of the original scope of a copy of the Blog index, after all! You've been a great help either way, so thanks a lot!

File: content/docs/command-reference/import.md (hunk 2 above)
Comment (jorgeorpinel) [reply content/docs/command-reference/import.md:172]: Looks like an accidental commit as he was also working on #1339, now merged. @utkarshsingh99 please just revert this whole file change from this branch.

**Comment by jorgeorpinel** [comment]
Some questions from me:

> The link on every blog post for the author's social media handle has been replaced with the link for their author page.

Should author names directly in the blog home page link to their pages too?
UPDATE: I see roger also noted this in https://github.com/iterative/dvc.org/pull/1364#issuecomment-634920420. I vote for yes.

> the authors node in graphql does not have the description field unlike the markdown files of authors...
> what do we want for descriptions? We could have something like longDescription on blog posts ... or we could use the Author's Markdown body content

Sorry, I'm not understanding this question. I see the author md content as desc already:

image

What's wrong with that?

> I'm open to rolling #1349 into this

@rogermparent if you're taking over I say go for it but if @utkarshsingh99 will be able to continue being the main contributor here I'd recommend we focus on a single issue for now and open a separate PR for that.

> GitHub didn't pick up on the deploy I triggered

It did after a while. Heroku probably only fires web hooks every several minutes 🙂

**Comment by rogermparent** [comment]
> Sorry, I'm not understanding this question. I see the author md content as desc already:
> 
> What's wrong with that?

That comment was made before I added the md content as descriptions. I think this method for descriptions is good to move forward with, especially where it's currently how the content is shaped. We can change it later if we decide to make bigger descriptive Author pages.

> > I'm open to rolling #1349 into this
> 
> @rogermparent if you're taking over I say go for it but if @utkarshsingh99 will be able to continue being the main contributor here I'd recommend we focus on a single issue for now and open a separate PR for that.

I actually already did it! That issue at its minimum was actually quite small, and the difference between having an array of multiple links and one link string is huge when designing the template. I'll certainly have to build on it though for features like icons, so I can do that in another issue and let @utkarshsingh99 continue on this one.

If I finish that before this PR is done, we can always merge the new schema and work it into the page here!

**Comment by jorgeorpinel** [comment]
> Design work for the author page header

Yeah it also has a weird responsive jump from 990px width to 1005px. It should have a max-height/scroll too probably? I was able to do this from DevTools:

image

- [ ] I would add the whole author list needs redesign. I suggest a vertical list with small rows (covering the full width, instead of 3-per-column) per post: the img to the left, and the gray area with meta info in the remaining space. I see author pages more like an archives so no need to be fancy with the design.

> Would you like to add links to other places where Author names show up, iterate on the design of the template?

This is a great suggestion for @utkarshsingh99 probably. But if the design stuff is becoming tedious or taking too long please feel free to move on to other issues or developing your GSoD project ideas, Utkarsh. Also if you have issues with Git since you're collaborating with someone else in the same PR, please let us know.

File: content/docs/command-reference/import.md (hunk 2 above)
Comment (utkarshsingh99) [reply content/docs/command-reference/import.md:172]: Yes, this was added accidentally. My bad

**Comment by utkarshsingh99** [comment]
Apologies for the late reply.
> Would you like to add links to other places where Author names show up,

Sure. I believe the only places that need to be linked are in the main blog-index which will have to be changed in FeedMeta component? @rogermparent 

> iterate on the design of the template?

I would definitely like to do it but I'm not particularly sure of what I have to do here. Does this mean going over the points listed in [#1364(comment)](https://github.com/iterative/dvc.org/pull/1364#issuecomment-634982031)

> But if the design stuff is becoming tedious or taking too long please feel free to move on to other issues or developing your GSoD project ideas

It's no problem. Working on this PR was fun and it would be great to help @rogermparent take this to the finish line.

>  if you have issues with Git since you're collaborating with someone else in the same PR, please let us know.

Sure. I believe I will have to remove the file mentioned in [#1364(comment)](https://github.com/iterative/dvc.org/pull/1364#discussion_r431482365) by reverting its commit specifically?

**Comment by rogermparent** [comment]
@utkarshsingh99 

> Apologies for the late reply.

No worries!

> Sure. I believe the only places that need to be linked are in the main blog-index which will have to be changed in FeedMeta component? @rogermparent

I believe that's the case! I don't see them anywhere else.

> I would definitely like to do it but I'm not particularly sure of what I have to do here. Does this mean going over the points listed in [#1364(comment)](https://github.com/iterative/dvc.org/pull/1364#issuecomment-634982031)

That's more or less the case. I just started on implementing the data sourcing for multiple links and icons, then overhauling pagination. I'll probably do these in other PRs that can be merged here later on.

If you'd like to copy the existing pagination logic for Blog feeds to get started, you're more than welcome to. The current blog pagination logic uses an unusual generator syntax so it could be difficult to understand, but it's readable enough to copy/paste if you want to get started on making the template pagination-compatible before I get to re-doing pagination.

You could also hard-code and/or define dummy data in consts, just for now while making a design, and then we'll plug real data into it later on when the data sourcing mechanism is done.

> Sure. I believe I will have to remove the file mentioned in [#1364(comment)](https://github.com/iterative/dvc.org/pull/1364#discussion_r431482365) by reverting its commit specifically?

I think we'd prefer you just revert with another commit. From what I understand, we have a bot that squashes branches before merge. In this case, we don't have to worry about history and just pushing a new commit that removes the file has no consequences.

Just to confirm, am I correct that we squash merges @shcheklein?

> P.S.: What are we going to do about the codeclimate failed checks and its suggested optimizations? I believe keeping certain similar code sections would be inevitable due to their partial difference.

The way I treat codeclimate is to look at it to see if anything addressable jumps out, but otherwise not worry about it. codeclimate being satisfied is optional, despite how the red X looks on GitHub's UI!

**Comment by shcheklein** [comment]
> Just to confirm, am I correct that we squash merges @shcheklein?

yes, I do squash and merge. (primarily since we tend to have a few commits with some images)

**Comment by rogermparent** [comment]
#1375 builds on the quick commit I made here to add special metadata to author links. It's heavily connected to this issue, despite being separate, so I figured I should point it out in a comment here as well as there.

@utkarshsingh99 I could merge this in to whatever you have pushed here, or you could handle it on your own. Your call! It'll probably help a lot having something to design around, and the SocialIcon component will also be useful.

**Comment by rogermparent** [comment]
@utkarshsingh99 I actually got a little too excited and went and did the merge myself. I hope the new social icons help you, and more importantly that I haven't clobbered anything you were working on!

Also, a hint on getting links on the blog post indexes: we're using the same `FeedMeta` component everywhere for author names, it's all about getting the author's `slug` delivered to `FeedMeta` to get an author link there.

**Comment by utkarshsingh99** [comment]
@rogermparent 
> @utkarshsingh99 I actually got a little too excited and went and did the merge myself.

Haha, that's okay. I will get on the linking task asap.

About the pagination, since you will be already making a newer pagination logic, I was thinking I would wait for that commit and then add pagination compatibility to the author template, so we won't have to add/delete any unnecessary code altogether. But if you want me to add current logic to the pages so you can test the pagination, I could do it before, if you want. 

> yes, I do squash and merge. (primarily since we tend to have a few commits with some images)

Okay, I'll do this. Just in case if I have any problem, should I prefer asking these kinds of doubts on Discord?

**Comment by rogermparent** [comment]
> Haha, that's okay. I will get on the linking task asap.

Great! Thanks a ton!

> About the pagination, since you will be already making a newer pagination logic, I was thinking I would wait for that commit and then add pagination compatibility to the author template, so we won't have to add/delete any unnecessary code altogether. But if you want me to add current logic to the pages so you can test the pagination, I could do it before, if you want.

Sounds good to me! I'll have that done sometime tomorrow.

**Comment by rogermparent** [comment]
Something I noticed that I should also mention is that we may want to not show the Author line in post items in the Author feed since it's pretty redundant.

**Comment by rogermparent** [comment]
A good example (the only one right now) of multiple author links is [Dmitry's Author page](https://dvc-landing-author-page-afpkg5.herokuapp.com/authors/dmitry_petrov)

**Comment by utkarshsingh99** [comment]
- [x] I have reverted the commit of the `import.md` file which was accidentally added here.

- [x] Author name is now clickable at the Blog Home page

I have another idea, but I don't think it is relevant here, maybe. In the blog home page, only the Title and the Meta Image of each blog post is clickable. 
I think it would be great if we could make the entire component clickable, except the author link and twitter link. 
Now the only way I can think of doing this with the author link constraint is by adding an event listener around the Meta component and check if the event's target is not the author link or twitter icon. 
Should we do this?

**Comment by rogermparent** [comment]
@utkarshsingh99 Nice! I merged in an update of the social icons I added earlier. I already merged your edits to make sure they work together.

> we could make the entire component clickable to display the blog post, except the author link and twitter link.
> Now the only way I can think of doing this with the author link constraint is by adding an event listener around the Meta component and check if the event's target is not the author link or twitter icon.
> Should we do this?

I agree expanding the card component (I think it's `Feed/Item`?) click area would be a good thing, but I don't like how much the suggested implementation leans on JS to make a link, even if we gate it behind progressive enhancement. An HTML/CSS solution would be best here, and I can think of a couple we can try.

Given the card's design with all the metadata and alternate links at the bottom, I would suggest making the blog link wrap around the image, title, and description while having the `FeedMeta` outside to contain its own links.  
This also feeds into the "We show authors on cards on the pages that only have one author" issue- if we make this separation in the card, we can also make the `FeedMeta` component change depending on context (e.g. don't show authors on the author pages), making the card component more re-usable for this and future purposes where we may want a different `FeedMeta`.

**Comment by rogermparent** [comment]
Oh no, I meant to push the final social links branch and broke something. I'm sorry, hold on, I'll fix it!

**Comment by rogermparent** [comment]
@utkarshsingh99 I mentioned I'd do icons "tomorrow" two days ago, but the social icons ended up ballooning unexpectedly so I wanted to get them fixed and merged ASAP. Anyway, I'll do pagination now in this repo and then you should have all the data you need to make a killer Author page!

With how the Author links are handled now, we can work with additional metadata for them. For example, we already extract the username from Twitter, GitHub, and LinkedIn profiles and expose it via GraphQL in case you want to display a username beside an icon or something.

**Comment by rogermparent** [comment]
@utkarshsingh99 I just added pagination to the Author pages! The implementation code is a little ugly, and I haven't migrated the blog's pagination just yet, but this should be enough such that you can get started considering it in the design.

I definitely have to clean it up tomorrow, though.

**Comment by utkarshsingh99** [comment]
I'm sorry for being inactive due to the past few days. There was an issue with the internet connectivity due to some weather disturbances in Mumbai.
Please have a look at the new header in the author's page. 
>we can also make the FeedMeta component change depending on context

I agree with @rogermparent 's idea that we need to have a new FeedMeta component for author page since the author name on the card on author page looks pretty unnecessary.

>With how the Author links are handled now, we can work with additional metadata for them. For example, we already extract the username from Twitter, GitHub, and LinkedIn profiles

Thanks! I have used username of the author in the header.

@shcheklein @jorgeorpinel @rogermparent Please let me know if the new design is good and is there anything else in the PR that remains?

**Comment by jorgeorpinel** [comment]
> Please let me know if the new design is good and is there anything else in the PR that remains

TBH I'm not following this PR's details closely, I'll let Roger or Ivan answer that. Thanks!

**Comment by shcheklein** [comment]
@utkarshsingh99 @rogermparent great work! It is definitely a good progress.

> is there anything else in the PR that remains?

Please check this comment:

https://github.com/iterative/dvc.org/pull/1364#issuecomment-634982031

> Please let me know if the new design is good 

Design looks good, way better than it was before, but we still need to involve our designer.

**Comment by utkarshsingh99** [comment]
> Authors (name and avatar) should be clickable on the main page as well as blog page.

@shcheklein done. 
@rogermparent Please check the code quality of the latest commit in FeedMeta, since I had to make some necessary omissions. For some reason, the Link wasn't working wrapped around the gatsby Image component until I removed the className attribute from Link. Might be because of the `@mixin link` property?

> Include author descriptions.

@shcheklein I think this has also been completed in [#1364(comment)](https://github.com/iterative/dvc.org/pull/1364#issuecomment-634982031). The author header does show their descriptions.

> Design looks good, way better than it was before, but we still need to involve our designer.

Thanks! Sure, I would be happy to make further changes.

**Comment by shcheklein** [comment]
- [ ] Something funny is happening here `marija-ili%C4%87-65b8a53` - https://dvc-landing-author-page-afpkg5.herokuapp.com/authors/marija_ilic
- [ ] Need to update @SvetaGr description here - https://dvc-landing-author-page-afpkg5.herokuapp.com/authors/svetlana_grinchenko
- [ ] Need to support basic Markdown in descriptions . E.g. to put creator of [DVC](https://dvc.org) instead of dvc.org. Check @dmpetrov ' description
- [ ] When author has multiple social network links, spacing is wrong:

Screen Shot 2020-06-06 at 2 54 05 PM

**Comment by utkarshsingh99** [comment]
@shcheklein I just checked.
> Something funny is happening here marija-ili%C4%87-65b8a53

This username, is coming from the backend itself in this way. We will have to probably make changes to the encoding of the characters, and I think something along this line might be the solution to all the 3 problems, but on the backend (my guess). I'm not sure of the backend code here, though.

> When author has multiple social network links, spacing is wrong

Actually, even the sizes of the icons have changed. I believe @rogermparent made some wilful changes in the previous commits for this? Not sure. 
I think if you review the latest code in SocialIcons, I can change that. (There is a `username` prop that I have added in the component, and it is undefined when called in the FeedMeta (not an error). Leaving it like that doesn't make any changes to the design though, just makes the component resuable). I was just curious if you guys thought it's a good practice or not.
But I still don't think that's the cause for this spacing error.

**Comment by utkarshsingh99** [comment]
So should I make a new component to render social icons in Author Header?
That should resolve this problem here.
`SocialIconsAuthorHeader`

What do you think about the other 3 problems though? @rogermparent

**Comment by rogermparent** [comment]
> Something funny is happening here marija-ili%C4%87-65b8a53 - https://dvc-landing-author-page-afpkg5.herokuapp.com/authors/marija_ilic

Ah, yeah the URL looked exactly like that at the time when I copy/pasted, something auto-encoded the URL on paste. The automatic behavior with a link string can be fixed in schema, but a quick fix could also include the explicit object syntax for Author links, using the encoded URL for the URL and the proper unicode name as the username. The format for specifying this object is the exact same as how the GraphQL objects look, just in YML.

> Need to update @SvetaGr description here - https://dvc-landing-author-page-afpkg5.herokuapp.com/authors/svetlana_grinchenko

I can't help here, I don't know what it needs to be updated to.

> Need to support basic Markdown in descriptions . E.g. to put creator of DVC instead of dvc.org. Check @dmpetrov ' description

This can be done the same way `descriptionLong` is done in Blog posts, by parsing frontmatter Markdown in the Model at build-time.

**Comment by jorgeorpinel** [comment]
> What do you think about the other 3 problems though?

Hey guys any way this PR can be separated or some aspects extracted into another issue so Utkarsh can finish his contribution and this gets merged? Trying to make it a little easier on him and also he's helping us in parallel with some content improvements that currently are of higher priority.

Thanks

**Comment by rogermparent** [comment]
@jorgeorpinel @utkarshsingh99 @shcheklein 

> Hey guys any way this PR can be separated or some aspects extracted into another issue so Utkarsh can finish his contribution and this gets merged?

From what I understand, we still need our designer to take a look at it and come up with something. I think this PR is in a good enough state such that once the designer is free and gives us his take, we can match this PR to that design.

We're probably not going to want to merge this PR before we get a design, none of its parts other than the social icons make sense standalone, and the social links are already in master, so I think we're already in a good state to leave this PR as it is and revisit it later.

**Comment by jorgeorpinel** [comment]
OK the last activity here was almost a month ago. As there doesn't seem to be a plan for the design, should we close this for now?

**Comment by jorgeorpinel** [comment]
(and keep the branch)

**Comment by shcheklein** [comment]
@jorgeorpinel reassigned to myself. I'll keep track of it. We got distracted by CML on the front-end dev part, but we can go back.

**Comment by jorgeorpinel** [comment]
OK I think it's safe to close this as stale at this point 🙂

**Comment by shcheklein** [comment]
@jorgeorpinel I assigned it to myself. Let's keep it open. I keep track of it.

**Comment by jorgeorpinel** [comment]
OK sure. I forgot you took it over.

**Comment by umnikovas** [comment]
[Design in Figma](https://www.figma.com/file/mbuTynd5nzDrTzpWnP2KgQ/DVC-Blog-page?node-id=0%3A1)

Overview:
Desktop
image
Mobile
image

We will need to upload better images for avatar pics. In the design, they are 100x100px.
Author descriptions were mentioned in this thread earlier. 
But probably at the first stage, not all contributors will have one:
image


In desktop social links are the same as in article items. 
In mobile the tap target should be 32x32 at least 
image

The rest of the page including pagination was made of reusable components as in [here](https://dvc.org/blog)
The major difference is that it has 9 article items of the same size. 
image

**Comment by shcheklein** [comment]
@umnikovas excellent stuff! just a few comments:

- not exactly clear how do we position social icons on mobile - what is the space between them?
- what happens if name is too long to fit into single line?
- on mobile text looks a bit too heavy - only a few pixels of the first post are visible? Not sure if it's a problem or not.
- descriptions might have links- we should account for that I guess. May be some other markup?

**Comment by umnikovas** [comment]
> * not exactly clear how do we position social icons on mobile - what is the space between them?

Looks like we have a 5px module in spacing.
I kept that system, so we have AuthorName (15px margin) (Icon 32px box)(5px margin)(Icon 32px box)

If there is not enough space for the social block to fit in the same line with the authorName it should wrap to the 2nd line.
The line-height of the Title is big, so there is only a need of a 5px margin-bottom between it and the icon block. And the same 5px margin-top for the description text.
image

  

> * what happens if name is too long to fit into single line?


If its this case as in Wikipedia, we will have to add `&shy;` in the string. Or maybe there are some more elegant solutions. 
image
image

> * on mobile text looks a bit too heavy - only a few pixels of the first post are visible? Not sure if it's a problem or not.

We can add a spoiler for that only in mobile. So the user can scroll for the blog posts faster. And only read about the author if he needs to, by opening up the spoiler. If the spoiler will be only on mobile, probably the description text can still be used in SEO. I think this could be a good thing to do, as the description could be content-rich.
image

But still, only some part of the first article will fit. There was a suggestion in this thread, that we can try to visualize the cards differently on the author page. Make them smaller. But I think it should be tested. As I don't think users have problems in scrolling.

> * descriptions might have links- we should account for that I guess. Maybe some other markup?

Don't know how good is an idea to put links in the author description, but we can do that I think: 
image

Can't say what type of markup could be useful in the author description. 
Perfectly if the author's description is short and concise.

**Comment by jorgeorpinel** [comment]
Ping @rogermparent or @julieg18 would it be possible for this to get taken over? Or should we close for now (turn into issue)? Thanks

**Comment by jorgeorpinel** [comment]
> would it be possible for this to get taken over?

False alarm, please disregard for now.

**Comment by shcheklein** [comment]
We already have #1120 . Closing this for now.

**Pull request closed without merging**